@ExportedService annotation (DS 1.4) not processed correctly by bnd 4.1.0

When trying to use the newly added @ExportedService annotation from DS 1.4 with bnd-maven-plugin 4.1.0 in order to declare the service.exported.interfaces property as follows:

```
@Component(service = TestService.class)
@ExportedService(service_exported_interfaces = TestService.class)
public class TestServiceImpl implements TestService {}
```

Results in the following error being reported by bnd:

```
[ERROR] Failed to execute goal biz.aQute.bnd:bnd-maven-plugin:4.1.0:bnd-process (default) on project test-project: Invalid data type ConcreteRef -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal biz.aQute.bnd:bnd-maven-plugin:4.1.0:bnd-process (default) on project test-project: Invalid data type ConcreteRef
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:213)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:51)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:309)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:194)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:107)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:955)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:564)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoFailureException: Invalid data type ConcreteRef
    at aQute.bnd.maven.plugin.BndMavenPlugin.reportErrorsAndWarnings (BndMavenPlugin.java:468)
    at aQute.bnd.maven.plugin.BndMavenPlugin.execute (BndMavenPlugin.java:328)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:134)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:208)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:51)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:309)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:194)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:107)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:955)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:564)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
```

--- a/biz.aQute.bndlib.tests/test/test/component_extra/DSPropertyAnnotationsTest.java
+++ b/biz.aQute.bndlib.tests/test/test/component_extra/DSPropertyAnnotationsTest.java
@@ -268,6 +268,64 @@ public class DSPropertyAnnotationsTest extends BndTestCase {
 	@ComponentPropertyType
 	@Retention(RetentionPolicy.CLASS)
 	@Target(ElementType.TYPE)
+	public static @interface IntArrayProp {
+		int[] value();
+	}
+
+	@IntArrayProp({
+		1, 2, 3
+	})
+	@Component
+	public static class IntArrayPropertyAnnotated {}
+
+	public void testIntArrayPropertyAnnotation() throws Exception {
+
+		Resource r = jar.getResource("OSGI-INF/" + IntArrayPropertyAnnotated.class.getName() + ".xml");
+
+		System.err.println(Processor.join(jar.getResources()
+			.keySet(), "\n"));
+		assertNotNull(r);
+		r.write(System.err);
+		XmlTester xt = new XmlTester(r.openInputStream(), "scr", "http://www.osgi.org/xmlns/scr/v1.3.0");
+		// Test the defaults
+		xt.assertAttribute(IntArrayPropertyAnnotated.class.getName(), "scr:component/implementation/@class");
+		xt.assertCount(1, "scr:component/property");
+		xt.assertAttribute("Integer", "scr:component/property[@name='int.array.prop']/@type");
+		xt.assertAttribute("1\n2\n3", "scr:component/property[@name='int.array.prop']");
+	}
+
+	@ComponentPropertyType
+	@Retention(RetentionPolicy.CLASS)
+	@Target(ElementType.TYPE)
+	public static @interface DefaultIntArrayProp {
+		int[] value() default {
+			1, 2, 3
+		};
+	}
+
+	@DefaultIntArrayProp
+	@Component
+	public static class DefaultIntArrayPropertyAnnotated {}
+
+	public void testDefaultIntArrayPropertyAnnotation() throws Exception {
+
+		Resource r = jar.getResource("OSGI-INF/" + DefaultIntArrayPropertyAnnotated.class.getName() + ".xml");
+
+		System.err.println(Processor.join(jar.getResources()
+			.keySet(), "\n"));
+		assertNotNull(r);
+		r.write(System.err);
+		XmlTester xt = new XmlTester(r.openInputStream(), "scr", "http://www.osgi.org/xmlns/scr/v1.3.0");
+		// Test the defaults
+		xt.assertAttribute(DefaultIntArrayPropertyAnnotated.class.getName(), "scr:component/implementation/@class");
+		xt.assertCount(1, "scr:component/property");
+		xt.assertAttribute("Integer", "scr:component/property[@name='default.int.array.prop']/@type");
+		xt.assertAttribute("1\n2\n3", "scr:component/property[@name='default.int.array.prop']");
+	}
+
+	@ComponentPropertyType
+	@Retention(RetentionPolicy.CLASS)
+	@Target(ElementType.TYPE)
 	public static @interface ClassArrayProp {
 		Class<?>[] value();
 	}
@@ -293,6 +351,24 @@ public class DSPropertyAnnotationsTest extends BndTestCase {
 		xt.assertAttribute("String", "scr:component/property[@name='class.array.prop']/@type");
 		xt.assertAttribute(List.class.getName() + "\n" + Set.class.getName() + "\n" + Queue.class.getName(),
 			"scr:component/property[@name='class.array.prop']");
+	}
+
+	@ClassArrayProp({})
+	@Component
+	public static class EmptyClassArrayPropertyAnnotated {}
+
+	public void testEmptyClassArrayPropertyAnnotation() throws Exception {
+
+		Resource r = jar.getResource("OSGI-INF/" + EmptyClassArrayPropertyAnnotated.class.getName() + ".xml");
+
+		System.err.println(Processor.join(jar.getResources()
+			.keySet(), "\n"));
+		assertNotNull(r);
+		r.write(System.err);
+		XmlTester xt = new XmlTester(r.openInputStream(), "scr", "http://www.osgi.org/xmlns/scr/v1.3.0");
+		// Test the defaults
+		xt.assertAttribute(EmptyClassArrayPropertyAnnotated.class.getName(), "scr:component/implementation/@class");
+		xt.assertCount(0, "scr:component/property");
 	}
 
 	@ComponentPropertyType

--- a/biz.aQute.bndlib.tests/test/test/component_extra/DSPropertyAnnotationsTest.java
+++ b/biz.aQute.bndlib.tests/test/test/component_extra/DSPropertyAnnotationsTest.java
@@ -6,6 +6,9 @@ import java.lang.annotation.ElementType;
 import java.lang.annotation.Retention;
 import java.lang.annotation.RetentionPolicy;
 import java.lang.annotation.Target;
+import java.util.List;
+import java.util.Queue;
+import java.util.Set;
 
 import org.osgi.service.component.annotations.Component;
 import org.osgi.service.component.annotations.ComponentPropertyType;
@@ -149,6 +152,33 @@ public class DSPropertyAnnotationsTest extends BndTestCase {
 	@ComponentPropertyType
 	@Retention(RetentionPolicy.CLASS)
 	@Target(ElementType.TYPE)
+	public static @interface MyClassProp {
+		Class<?> value();
+	}
+
+	@MyClassProp(List.class)
+	@Component
+	public static class ClassDSPropertyAnnotated {}
+
+	public void testClassPropertyAnnotation() throws Exception {
+
+		Resource r = jar.getResource("OSGI-INF/" + ClassDSPropertyAnnotated.class.getName() + ".xml");
+
+		System.err.println(Processor.join(jar.getResources()
+			.keySet(), "\n"));
+		assertNotNull(r);
+		r.write(System.err);
+		XmlTester xt = new XmlTester(r.openInputStream(), "scr", "http://www.osgi.org/xmlns/scr/v1.3.0");
+		// Test the defaults
+		xt.assertAttribute(ClassDSPropertyAnnotated.class.getName(), "scr:component/implementation/@class");
+		xt.assertCount(1, "scr:component/property");
+		xt.assertAttribute(List.class.getName(), "scr:component/property[@name='my.class.prop']/@value");
+		xt.assertAttribute("String", "scr:component/property[@name='my.class.prop']/@type");
+	}
+
+	@ComponentPropertyType
+	@Retention(RetentionPolicy.CLASS)
+	@Target(ElementType.TYPE)
 	public static @interface MyMultiProp {
 		String a_string_prop();
 
@@ -233,6 +263,66 @@ public class DSPropertyAnnotationsTest extends BndTestCase {
 		xt.assertCount(1, "scr:component/property");
 		xt.assertAttribute("String", "scr:component/property[@name='array.prop']/@type");
 		xt.assertAttribute("fizz\nbuzz\nfizzbuzz", "scr:component/property[@name='array.prop']");
+	}
+
+	@ComponentPropertyType
+	@Retention(RetentionPolicy.CLASS)
+	@Target(ElementType.TYPE)
+	public static @interface ClassArrayProp {
+		Class<?>[] value();
+	}
+
+	@ClassArrayProp({
+		List.class, Set.class, Queue.class
+	})
+	@Component
+	public static class ClassArrayPropertyAnnotated {}
+
+	public void testClassArrayPropertyAnnotation() throws Exception {
+
+		Resource r = jar.getResource("OSGI-INF/" + ClassArrayPropertyAnnotated.class.getName() + ".xml");
+
+		System.err.println(Processor.join(jar.getResources()
+			.keySet(), "\n"));
+		assertNotNull(r);
+		r.write(System.err);
+		XmlTester xt = new XmlTester(r.openInputStream(), "scr", "http://www.osgi.org/xmlns/scr/v1.3.0");
+		// Test the defaults
+		xt.assertAttribute(ClassArrayPropertyAnnotated.class.getName(), "scr:component/implementation/@class");
+		xt.assertCount(1, "scr:component/property");
+		xt.assertAttribute("String", "scr:component/property[@name='class.array.prop']/@type");
+		xt.assertAttribute(List.class.getName() + "\n" + Set.class.getName() + "\n" + Queue.class.getName(),
+			"scr:component/property[@name='class.array.prop']");
+	}
+
+	@ComponentPropertyType
+	@Retention(RetentionPolicy.CLASS)
+	@Target(ElementType.TYPE)
+	public static @interface DefaultClassArrayProp {
+		Class<?>[] value() default {
+			List.class, Set.class, Queue.class
+		};
+	}
+
+	@DefaultClassArrayProp
+	@Component
+	public static class DefaultClassArrayPropertyAnnotated {}
+
+	public void testDefaultClassArrayPropertyAnnotation() throws Exception {
+
+		Resource r = jar.getResource("OSGI-INF/" + DefaultClassArrayPropertyAnnotated.class.getName() + ".xml");
+
+		System.err.println(Processor.join(jar.getResources()
+			.keySet(), "\n"));
+		assertNotNull(r);
+		r.write(System.err);
+		XmlTester xt = new XmlTester(r.openInputStream(), "scr", "http://www.osgi.org/xmlns/scr/v1.3.0");
+		// Test the defaults
+		xt.assertAttribute(DefaultClassArrayPropertyAnnotated.class.getName(), "scr:component/implementation/@class");
+		xt.assertCount(1, "scr:component/property");
+		xt.assertAttribute("String", "scr:component/property[@name='default.class.array.prop']/@type");
+		xt.assertAttribute(List.class.getName() + "\n" + Set.class.getName() + "\n" + Queue.class.getName(),
+			"scr:component/property[@name='default.class.array.prop']");
 	}
 
 	@ComponentPropertyType

--- a/biz.aQute.bndlib/src/aQute/bnd/component/DSAnnotationReader.java
+++ b/biz.aQute.bndlib/src/aQute/bnd/component/DSAnnotationReader.java
@@ -633,7 +633,9 @@ public class DSAnnotationReader extends ClassDataCollector {
 			for (Entry<String, Object> entry : componentPropertyAnnotation.entrySet()) {
 				String key = entry.getKey();
 				Object value = entry.getValue();
-				handleValue(key, value, value instanceof TypeRef, null);
+				boolean isClass = value instanceof TypeRef || (value.getClass()
+					.isArray() && ((Object[]) value)[0] instanceof TypeRef);
+				handleValue(key, value, isClass, null);
 			}
 		}
 

--- a/biz.aQute.bndlib/src/aQute/bnd/component/DSAnnotationReader.java
+++ b/biz.aQute.bndlib/src/aQute/bnd/component/DSAnnotationReader.java
@@ -633,8 +633,25 @@ public class DSAnnotationReader extends ClassDataCollector {
 			for (Entry<String, Object> entry : componentPropertyAnnotation.entrySet()) {
 				String key = entry.getKey();
 				Object value = entry.getValue();
-				boolean isClass = value instanceof TypeRef || (value.getClass()
-					.isArray() && ((Object[]) value)[0] instanceof TypeRef);
+				boolean isClass;
+				if (value instanceof TypeRef) {
+					isClass = true;
+				} else if (value.getClass()
+					.isArray()) {
+					Object[] valueArray = (Object[]) value;
+					if (valueArray.length == 0) {
+						// If there are no values then it doesn't matter whether
+						// this
+						// is a class or not as we pass null for the type and it
+						// will
+						// eventually get interpreted as a string
+						isClass = false;
+					} else {
+						isClass = valueArray[0] instanceof TypeRef;
+					}
+				} else {
+					isClass = false;
+				}
 				handleValue(key, value, isClass, null);
 			}
 		}

